### ISSUE TITLE
Idea: Using Query Builder / ORM for database access like SQLAlchemy or Peewee

### ISSUE THREAD

**Issue by petercsiba** [AtticusZeller/fastapi_supabase_template#137]
**Is your feature request related to a problem? Please describe.**
Thanks for putting this together!

One thing which annoys me about Supabase-Py is the lack of Query Builder / ORM which feels like makes the queries fragile OR you have to have a full CRUD test for it. 

**Describe the solution you'd like**
Did you consider generating models out of the Postgres Database with say [sqlacodegen](https://pypi.org/project/sqlacodegen/)?
I.e. instead of doing
```
supabase.table("countries").select("*").eq("country", "IL").execute()
```
you could with SQLAlchemy or Peewee ORMs do like:
```
Countries.get(country == "IL")
```
which catches query problems at build time.

**Describe alternatives you've considered**
Current solution feels good enough. 

**Additional context**
I have started executing on this idea with Peewee, but it ain't yet fully production ready. Just wanted to give you an idea
https://github.com/petercsiba/supawee?tab=readme-ov-file#example-models

**Comment by Atticuszz** [comment]
Sorry to have u waiting for few weeks! busy for the bottom of the year. i am very appreciated for your instructive suggestion.

> One thing which annoys me about Supabase-Py is the lack of Query Builder / ORM which feels like makes the queries fragile OR you have to have a full CRUD test for it.

you are definitely right dude! i  did  find the discussion about `ORM` in the `supabase-py` issues, and looks no plan for integrating supabase with `ORM` style query, but i think it deserves a try. causes i genuinely prefer the object-oriented mode, and i thinks it should have a plan for it.

>  this idea with Peewee
the idea will help us reduce too much time on writing ORM model manually.sounds great! 
[tiangolo](https://github.com/tiangolo) who created the `FastAPI` also build the `ORM` tools called [sqlmodel](https://github.com/fastapi/sqlmodel) , it looks like a good high-level abstraction for SQLAlchemy with pydantic so i think this one maybe suitable for fastapi, [sqlacodegen](https://pypi.org/project/sqlacodegen/) suit for it i think

in short, the following works:
- [ ] `SqlModel ORM ` instead  of `supabase.table("countries").select("*").eq("country", "IL").execute()`
- [ ] sqlacodegen for sqlmodel

your ideas do help a lot!

**Comment by petercsiba** [comment]
NP thanks for the thoughtful response and research Atticuszz!

Good point +1, yes for a Fastapi library like yours sqlmodel w/ sqlacodegen feels like the best integration,
since it is most likely to match "vibes" of FastAPI.

(Just rambling lol) 
Was thinking if adding `sqlmodel` to this template si too "opinionated",
but then this is a template, so from a projects lifespan perspective forking this template might be the best point-of-time to commit to a way to interact with the database. 

That said, this is just an idea, thanks again for putting this repo together!

**Comment by Atticuszz** [comment]
Hey, thanks for bringing this up. After thinking it over, I'm leaning towards implementing SQLModel. Here's why:

We could use Supabase primarily as a database UI tool, while handling database operations through a more universal approach like SQLModel. This gives us much more flexibility and portability. It avoids tying us too closely to Supabase-specific implementations, which could be limiting in the long run.

SQLModel offers the best of both worlds - we can use object-oriented queries or fall back to SQL-like syntax when needed. This versatility is key for a template that needs to cater to various project requirements.

The project's roadmap is promising too, with async support coming up. This forward-thinking approach aligns well with our need for a future-proof solution.

What are your thoughts on this approach? It seems to strike a good balance between providing a solid starting point and maintaining flexibility for different project needs.

**Comment by AtticusZeller** [comment]
it starts on #143 now ,coming soon together with self-host supabase  and sqlmodel orm 😆

**Comment by AtticusZeller** [comment]
merged #143
